File: 10487 Closest Sums/solution.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b) {
    return *(int*)a > *(int*)b;
}

int main() {
    int cases = 0, n;
    while(scanf("%d", &n) && n != 0) {
        printf("Case %d:\n", ++cases);
        int set[n];
        for(int i = 0; i < n; ++i)
            scanf("%d", &set[i]);
        qsort(set, n, sizeof(int), compare);
        int m;
        scanf("%d", &m);
        while(m--) {
            int query;
            scanf("%d", &query);
            int ans = set[0] + set[1];
            for(int i = 0; i < n - 1 && set[i] <= query; ++i) {
                for(int j = i + 1; j < n; ++j) {
                    int sum = set[i] + set[j];
                    ans = abs(query - ans) < abs(query - sum) ? ans : sum;
                    if(sum > query)
                        break;
                }
            }
            printf("Closest sum to %d is %d.\n", query, ans);
        }
    }
	return 0;
}

File: 10487 Closest Sums/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    StreamTokenizer st = new StreamTokenizer(System.in);
        int cases = 0, n;
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0) {
            output.append("Case " + (++cases) + ":\n");
            int[] set = new int[n];
            for(int i = 0; i < n; ++i) {
                st.nextToken();
                set[i] = (int)st.nval;
            }
            Arrays.sort(set);
            st.nextToken();
            int m = (int)st.nval;
            while(m-- > 0) {
                st.nextToken();
                int query = (int)st.nval, ans = set[0] + set[1];
                for(int i = 0; i < n - 1 && set[i] <= query; ++i) {
                    for(int j = i + 1; j < n; ++j) {
                        int sum = set[i] + set[j];
                        ans = Math.abs(query - ans) < Math.abs(query - sum) ? ans : sum;
                        if(sum > query)
                            break;
                    }
                }
                output.append("Closest sum to ").append(query).append(" is ").append(ans).append(".\n");
            }
    	}
        System.out.print(output);
	}
}
